File: MyLeetCode_Go/study/LeetCode/Problem_146_LRUCache/Problem_146_LRUCache.go
package main

import "fmt"

// 本题测试链接 : https://leetcode.com/problems/lru-cache/

type LRUCache struct {
	cache MyCache
}

func Constructor(capacity int) LRUCache {
	return LRUCache{cache: NewMyCache(capacity)}
}

func (this *LRUCache) Get(key int) int {
	ans := this.cache.Get(key)
	if ans == nil {
		return -1
	} else {
		return ans.(int)
	}
}

func (this *LRUCache) Put(key int, value int) {
	this.cache.Set(key, value)
}

type Node struct {
	Key   interface{}
	Value interface{}
	Last  *Node
	Next  *Node
}

func NewNode(key, value interface{}) *Node {
	return &Node{Key: key, Value: value}
}

type NodeDoubleLinkedList struct {
	Head *Node
	Tail *Node
}

func NewNodeDoubleLinkedList() *NodeDoubleLinkedList {
	return &NodeDoubleLinkedList{}
}

// 现在来了一个新的node，请挂到尾巴上去
func (linkedlist *NodeDoubleLinkedList) AddNode(newNode *Node) {
	if newNode == nil {
		return
	}
	if linkedlist.Head == nil {
		linkedlist.Head = newNode
		linkedlist.Tail = newNode
	} else {
		linkedlist.Tail.Next = newNode
		newNode.Last = linkedlist.Tail
		linkedlist.Tail = newNode
	}
}

// node 入参，一定保证！node在双向链表里！
// node原始的位置，左右重新连好，然后把node分离出来
// 挂到整个链表的尾巴上
func (linkedlist *NodeDoubleLinkedList) moveNodeToTail(node *Node) {
	if linkedlist.Tail == node {
		return
	}
	if linkedlist.Head == node {
		linkedlist.Head = node.Next
		linkedlist.Head.Last = nil
	} else {
		node.Last.Next = node.Next
		node.Next.Last = node.Last
	}
	node.Last = linkedlist.Tail
	node.Next = nil
	linkedlist.Tail.Next = node
	linkedlist.Tail = node
}

func (linkedlist *NodeDoubleLinkedList) RemoveHead() *Node {
	if linkedlist.Head == nil {
		return nil
	}
	res := linkedlist.Head
	if linkedlist.Head == linkedlist.Tail {
		linkedlist.Head = nil
		linkedlist.Tail = nil
	} else {
		linkedlist.Head = res.Next
		res.Next = nil
		linkedlist.Head.Last = nil
	}
	return res
}

type MyCache struct {
	KeyNodeMap map[interface{}]interface{}
	NodeList   *NodeDoubleLinkedList
	capacity   int
}

func NewMyCache(cap int) MyCache {
	return MyCache{
		KeyNodeMap: make(map[interface{}]interface{}),
		NodeList:   NewNodeDoubleLinkedList(),
		capacity:   cap,
	}
}

func (mc *MyCache) Get(key interface{}) interface{} {
	if _, ok := mc.KeyNodeMap[key]; ok {
		res := mc.KeyNodeMap[key].(*Node)
		mc.NodeList.moveNodeToTail(res)
		return res.Value
	}
	return nil
}

// Set(Key, Value)
// 新增  更新value的操作
func (mc *MyCache) Set(key, value interface{}) {
	if _, ok := mc.KeyNodeMap[key]; ok {
		node := mc.KeyNodeMap[key].(*Node)
		node.Value = value
		mc.NodeList.moveNodeToTail(node)
	} else { // 新增！
		newNode := NewNode(key, value)
		mc.KeyNodeMap[key] = newNode
		mc.NodeList.AddNode(newNode)
		if len(mc.KeyNodeMap) == mc.capacity+1 {
			mc.removeMostUnusedCache()
		}
	}
}

func (mc *MyCache) removeMostUnusedCache() {
	removeNode := mc.NodeList.RemoveHead()
	delete(mc.KeyNodeMap, removeNode.Key)
}

func main() {
	lRUCache := Constructor(2)
	lRUCache.Put(1, 1)
	lRUCache.Put(2, 2)
	ans := lRUCache.Get(1)
	fmt.Println(ans)
	lRUCache.Put(3, 3)
	ans = lRUCache.Get(2)
	fmt.Println(ans)
	lRUCache.Put(4, 4)
	ans = lRUCache.Get(1)
	fmt.Println(ans)
	ans = lRUCache.Get(3)
	fmt.Println(ans)
	ans = lRUCache.Get(4)
	fmt.Println(ans)
}

File: MyLeetCode_Java/src/main/java/_Simulation/Problem_146_LRUCache.java
package _Simulation;

import java.util.HashMap;
import java.util.Map;
import java.util.TreeMap;

public class Problem_146_LRUCache {

    static class LRUCache {

        private static class Node {
            private int key; // NOTE: 从map删除头结点的时候使用
            private int val;
            private Node pre;
            private Node next;

            public Node(int k, int v) {
                key = k;
                val = v;
            }
        }

        private static class DoubleLinkedList {
            private Node head, tail;

            private void addNode(Node node) {
                if (head == null) {
                    head = node;
                    tail = node;
                    return;
                }
                tail.next = node;
                node.pre = tail;
                tail = node;
            }

            private int removeHead() {
                if (head == null) return -1;
                int res = head.key;
                if (head == tail) {
                    head = null;
                    tail = null;
                    return res;
                }
                head = head.next;
                head.pre = null;
                return res;
            }

            private void moveNodeToTail(Node node) {
                if (node == tail) return;
                if (node == head) {
                    head = head.next;
                    head.pre = null;
                } else {
                    Node pre = node.pre;
                    Node next = node.next;
                    pre.next = next;
                    next.pre = pre;
                }
                tail.next = node;
                node.pre = tail;
                tail = node;
            }
        }

        private Map<Integer, Node> map;
        private int cap;
        private DoubleLinkedList myCache;

        public LRUCache(int capacity) {
            cap = capacity;
            map = new HashMap<>();
            myCache = new DoubleLinkedList();
        }

        public int get(int key) {
            int res = -1;
            if (map.containsKey(key)) {
                Node node = map.get(key);
                res = node.val;
                myCache.moveNodeToTail(node);
            }
            return res;
        }

        public void put(int key, int value) {
            Node node = null;
            if (map.containsKey(key)) {
                node = map.get(key);
                node.val = value;
                myCache.moveNodeToTail(node);
            } else {
                if (map.size() == cap) {
                    int val = myCache.removeHead();
                    map.remove(val);
                }
                node = new Node(key, value);
                map.put(key, node);
                myCache.addNode(node);
            }
        }
    }

    public static void main(String[] args) {

        LRUCache lruCache = new LRUCache(2);
        lruCache.put(1, 0);
        lruCache.put(2, 2);
        int res = lruCache.get(1);
        System.out.println(res);
        lruCache.put(3, 3);
        res = lruCache.get(2);
        System.out.println(res);
        lruCache.put(4, 4);
        res = lruCache.get(1);
        System.out.println(res);
    }
}
